enhance(@graphql-hive/yoga): improvements with Yoga's features (#6118)
graphql-hive/console: .changeset/spotty-toes-help.md

---
'@graphql-hive/yoga': patch
---

> Align with the latest yoga;

Reviewer: link to "latest yoga" and maybe specify the version here as well.

Author: Updated, good point! Let me know what you think!

> Align with the latest yoga;

Author: ```suggestion
Align with GraphQL Yoga >=5.10.4 which has the relevant features mentioned below;
```

---
'@graphql-hive/yoga': patch
---

Align with the latest yoga;

- Remove \`tiny-lru\` dependency, and use `createLruCache` from `graphql-yoga`
- Use Explicit Resource Management of GraphQL Yoga plugins for disposal which already respect Node.js process termination
- Use \`context.waitUntil\` which is handled by the environment automatically, if not `@whatwg-node/server` already takes care of it with above.
- Respect the given `fetchAPI` by GraphQL Yoga(might be Hive Gateway) for the `fetch` function

> - Avoid using `async/await` for performance reasons, because in case of cache-hit, it returns synchronously

Author: ```suggestion
```

---
'@graphql-hive/yoga': patch
---

Align with GraphQL Yoga >=5.10.4 which has the relevant features mentioned below;

- Remove \`tiny-lru\` dependency, and use `createLruCache` from `graphql-yoga`

> - - Yoga already provides a LRU cache implementation, so we can use that instead of having a separate dependency

Reviewer 2: what is the difference between Yoga's implementation and `tiny-lru`?

Author: Yoga uses `lru-cache` package, and I don't think there is a big difference. My idea is to avoid introducing an extra dependency in the plugin but use the one Yoga provides to reduce the duplicate code.

Reviewer 2: if there is no difference, why does Yoga adds this to itself?

Author: In Yoga, we use LRU caching for parsing and validation caching, and we preferred `lru-cache` at some point. 
But this plugin introduces a new dependency `tiny-lru` which does the same job. In order to avoid introducing new dependencies doing the same job, and it is an unnecessary dependency especially for the bundle size. But instead, Yoga exports `createLRUCache` function for plugins to re-use the same thing Yoga uses to avoid extra dependencies, code etc.

Reviewer 2: yes I understand
I'm just asking, if the fact that now you expose this functionality means we need to maintain it and people will rely on it, and if so, is it really yoga's job?

Reviewer 2: how important is the bundle size here?

Author: > how important is the bundle size here?

In Yoga, we always avoided to introduce new dependencies, duplicate code etc for bundle size. Extra dependencies mean extra bundle size not just for Yoga but also for Gateway which has this package built-in.

> I'm just asking, if the fact that now you expose this functionality means we need to maintain it and people will rely on it, and if so, is it really yoga's job?

We already maintain LRU caching in Yoga. So I don't think it is too much work here. It is already there in Yoga and used by many plugins.

Author: For me any extra non-optimal thing is important because when they become together, it becomes an issue that we need to solve later by analyzing each of these. I just want to prevent them growing before.

> Align with GraphQL Yoga >=5.10.4 which has the relevant features mentioned below;
> 
> - Remove \`tiny-lru\` dependency, and use `createLruCache` from `graphql-yoga`
> - - Yoga already provides a LRU cache implementation, so we can use that instead of having a separate dependency
> - Use Explicit Resource Management of GraphQL Yoga plugins for disposal which already respect Node.js process termination
> - - The feature has been implemented in `@whatwg-node/server` which is used by GraphQL Yoga. [Learn more about this feature](https://github.com/ardatan/whatwg-node/pull/1830)
> - Use \`context.waitUntil\` which is handled by the environment automatically, if not `@whatwg-node/server` already takes care of it with above.
> - - The feature has been implemented in `@whatwg-node/server` which is used by GraphQL Yoga. [Learn more about this feature](
>     https://github.com/ardatan/whatwg-node/pull/1830)
> - Respect the given `fetchAPI` by GraphQL Yoga(might be Hive Gateway) for the `fetch` function
> - - Hive Gateway uses `fetchAPI` given to GraphQL Yoga for the entire Fetch API implementation, so if Hive Client respects that, we have a control over HTTP calls done internally by the gateway
> - Respect Yoga's \`logger\` for logging
> - - Similar to above, we want to respect the logger provided by Yoga to have a better control over logging

Reviewer 3: I tri
```suggestion
- Upgrade to `graphql-yoga` >= `5.10.4`
- Improve graceful process termination on Node.js by leveraging `graphql-yoga`'s [dispose lifecycle hooks](https://the-guild.dev/graphql/yoga-server/docs/features/envelop-plugins#ondispose)
- Improve Cloudflare Worker runtime support by registering pending usage reporting requests using the [`waitUntil` API](https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil)
- Use the logger instance from the `graphql-yoga` instance, for a more unified logging experience
- Use the `fetch` API implementation on the `graphql-yoga` instance for HTTP calls
- Replace the internal \`tiny-lru\` dependency with `graphql-yoga`'s internal LRU cache implementation
```

Author: Looks way better! Thanks 🙏🏻
